editor: make underscores and newlines subword boundaries

durian-axolotl opened the pull request:
Closes #2

Fixed two bugs in Zed's subword navigation system:

## Underscore Navigation

### Problem

- Forward navigation: Only detected transitions *to* underscores, causing movement to stop *before* the underscore
- Backward navigation: Only detected transitions *from* underscores, causing movement to stop *after* the underscore

This created inconsistent behavior where the underscore was grouped with different words depending on direction.

### Fix

Added bidirectional underscore boundary detection.

### Result

`^` marks cursor in the beginning and `|` afterwards.

- Before: For `this_feature`:
  - Forward: `this^_feature|` (skipped the underscore as a separate stop)
  - Backward: `|this_|feature^` (inconsistent grouping)
- After: For `this_feature`:
  - Forward: `this^_|feature|` (stops after `this`, after `_`, and after `feature`)
  - Backward: `|this|_|feature^` (stops before `this`, before `_`, and before `feature`)

The underscore is now treated consistently as a boundary on both sides.

## Newline Navigation

### Problem

Newlines were not treated as subword boundaries.

### Fix

When adjacent to a newline, navigation now stops at the other side of the newline first instead of jumping over it.

### Result

- Before: For `one_word↵other_word`:
  - Forward: `one_word^↵other|_word`
  - Backward: `one_|word↵^other_word`
- After: For `one_word↵other_word`:
  - Forward: `one_word^↵|other_word`
  - Backward: `one_word|↵^other_word`


---

Both fixes ensure consistent boundary detection in both forward and backward directions.

Release Notes:

- editor: make underscores and newlines subword boundaries

durian-axolotl committed:
editor: make underscores and newlines subword boundaries

ConradIrwin commented:
Thank you for this!

I'm a bit surprised that we should stop twice on an underscore, when testing this in RustRover (a Jetbrains IDE) I found that:

* four^ one^_two^_three^ in the forward direction it stops at the end of words and subwords
* ^four ^one_^two_^three in the backward direction it stops at the start of words and subwords

Which seems to be the same as we do today.

Should we add a setting to make option-left/right do this? Or maybe we can close the issue and explain how to use the existing keybindings in the keymap?

durian-axolotl commented:
You are correct with the underscore behavior in Jetbrains IDEs, I just tested it in PyCharm. I just implemented the underscore behavior how I understood it from the example in the issue. But since Zed's behavior is already consistent with the Jetbrains IDEs I don't think we should change it.

Regarding the newlines: in my tests Jetbrains IDEs respect newlines as subword boundaries, like how I implemented it. So my suggestion would be to revert the underscore behavior but keep the changed newline stuff. Is this something you would agree on?

For the issue, I think its sufficient to close it and explain the keybindings.

ConradIrwin commented:
Great, thanks!
